temporal-sdk/src/main/java/io/temporal/internal/worker/EagerActivitySlotsReservation.java

```diff
@@ -0,0 +1,88 @@
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.internal.worker;
+
+import io.temporal.api.command.v1.Command;
+import io.temporal.api.command.v1.ScheduleActivityTaskCommandAttributes;
+import io.temporal.api.enums.v1.CommandType;
+import io.temporal.api.workflowservice.v1.PollActivityTaskQueueResponse;
+import io.temporal.api.workflowservice.v1.RespondWorkflowTaskCompletedRequest;
+import io.temporal.api.workflowservice.v1.RespondWorkflowTaskCompletedResponse;
+
+public class EagerActivitySlotsReservation implements AutoCloseable {
+  private final EagerActivityDispatcher eagerActivityDispatcher;
+  private int acquiredReservationCount = 0;
+
+  EagerActivitySlotsReservation(
+      EagerActivityDispatcher eagerActivityDispatcher,
+      RespondWorkflowTaskCompletedRequest.Builder mutableRequest) {
+    this.eagerActivityDispatcher = eagerActivityDispatcher;
+    applyToRequest(mutableRequest);
+  }
+
+  public void applyToRequest(RespondWorkflowTaskCompletedRequest.Builder mutableRequest) {
+    for (int i = 0; i < mutableRequest.getCommandsCount(); i++) {
+      Command command = mutableRequest.getCommands(i);
+      if (command.getCommandType() != CommandType.COMMAND_TYPE_SCHEDULE_ACTIVITY_TASK) continue;
+
+      ScheduleActivityTaskCommandAttributes commandAttributes =
+          command.getScheduleActivityTaskCommandAttributes();
+      if (!commandAttributes.getRequestEagerExecution()) continue;
+
+      if (this.eagerActivityDispatcher.tryReserveActivitySlot(commandAttributes)) {
+        this.acquiredReservationCount++;
+      } else {
+        mutableRequest.setCommands(
+            i,
+            command.toBuilder()
+                .setScheduleActivityTaskCommandAttributes(
+                    commandAttributes.toBuilder().setRequestEagerExecution(false)));
+      }
+    }
+  }
+
+  public void handleResponse(RespondWorkflowTaskCompletedResponse serverResponse) {
+    if (serverResponse.getActivityTasksCount() > this.acquiredReservationCount)
+      throw new IllegalStateException(
+          String.format(
+              "Unexpectedly received %d eager activities though we only requested %d",
+              serverResponse.getActivityTasksCount(), this.acquiredReservationCount));
+
+    if (this.acquiredReservationCount == 0) return;
+
+    for (PollActivityTaskQueueResponse act : serverResponse.getActivityTasksList())
+      this.eagerActivityDispatcher.dispatchActivity(act);
+
+    this.eagerActivityDispatcher.releaseActivitySlotReservations(
```

Comment:
Probably won't affect much because the dispatch calls are fast (right?), but the logic in my head would say to move this above the dispatch step to give slots back to the activity worker as soon as possible